src/storage/owncloud.py

```diff
@@ -26,5 +26,48 @@ def upload(file: str) -> str:
 
     # upload file
     oc.put_file(remotefile, file)
-    return str(oc.share_file_with_link(remotefile).get_link())
+    return _transform_url(str(oc.share_file_with_link(remotefile).get_link()))
+
+
+def _login(oc: owncloud.Client):
+    kvstore_dirty = False
+
+    # initial case: no username or password present
+    if "username" not in kvstore.keys() or "password" not in kvstore.keys():
+        kvstore["username"], kvstore["password"] = _get_credentials()
+        kvstore_dirty = True
+
+    while True:
+        try:
+            # try to log in with given credentials
+            oc.login(kvstore["username"], kvstore["password"])
+
+            # login worked: save credentials if they're new, break loop
+            if kvstore_dirty:
+                kvstore.sync()
+            return True
+        except owncloud.HTTPResponseError as e:
+            if e.status_code == 401:  # auth failure: read credentials again
+                print("Authentication Failure.")
+                kvstore["username"], kvstore["password"] = _get_credentials()
+                kvstore_dirty = True
+            else:  # other owncloud error
+                logging.error("Upload failed. Owncloud Server reply: {}".format(e.status_code))
+                return False
+        except RuntimeError as e:  # any other error
+            logging.error("Upload failed. Error details: {}".format(e))
+            return False
+
+
+def _get_credentials():
+    from getpass import getpass
+
+    user = getpass("OwnCloud username:")
+    password = getpass("OwnCloud password:")
+
+    return user, password
+
 
+def _transform_url(url: str) -> str:
```

Comment:
OwnCloud does not support direct links to files because of concerns about cross site scripting. So for images, transforming the URL would be a solution. We would need the dimensions of the image though. Looking at multimedia support further complicating things, I'm starting to think we should just leave the link as is. That's how it's supposed to be shared anyways. What do you think?